### PR TITLE
Add dt_filesocket to sapmachine17

### PR DESCRIPTION
**Pull request by schmelter-sap**
This adds support for dt_filesocket to sapmachine17. In contrast to the sapmachine11 implementation, this uses Unix Domain Sockets on Windows too, since Java 16 added support for them.

fixes #1611

### CODE REVIEW

**Comment by SapMachine** [comment]
Hello @schmelter-sap, I'm sorry, this pull request doesn't meet the expectations. The pull request description has an invalid format. Please have a look at https://github.com/SAP/SapMachine/wiki/Formal-Requirements-of-Pull-Requests .

**Comment by SapMachine** [comment]
Hello @schmelter-sap, I'm sorry, this pull request doesn't meet the expectations. The pull request description has an invalid format. Please have a look at https://github.com/SAP/SapMachine/wiki/Formal-Requirements-of-Pull-Requests .

**Comment by SapMachine** [comment]
Hello @schmelter-sap, this pull request fulfills all formal requirements.

**Comment by SapMachine** [comment]
Hello @schmelter-sap, this pull request fulfills all formal requirements.

**Comment by SapMachine** [comment]
Hello @schmelter-sap, this pull request fulfills all formal requirements.

**Review by MBaesken — COMMENTED** [review]
Looks mostly good but added a couple of minor comments/suggestions/questions

File: make/common/modules/LauncherCommon.gmk [hunk 1]
```diff
@@ -1,5 +1,5 @@
 #
-# Copyright (c) 2011, 2020, Oracle and/or its affiliates. All rights reserved.
+# Copyright (c) 2011, 2024, Oracle and/or its affiliates. All rights reserved.
```
Comment (RealCLanger) [inline make/common/modules/LauncherCommon.gmk:2]: Do not change Oracle copyright here, would be an unnecessary diff.

File: make/modules/jdk.jdwp.agent/Lib.gmk [hunk 2]
```diff
@@ -1,5 +1,5 @@
 #
-# Copyright (c) 2011, 2020, Oracle and/or its affiliates. All rights reserved.
+# Copyright (c) 2011, 2024, Oracle and/or its affiliates. All rights reserved.
```
Comment (RealCLanger) [inline make/modules/jdk.jdwp.agent/Lib.gmk:2]: Same here, do not change Oracle copyright

File: src/jdk.jdwp.agent/share/native/libdt_filesocket/fileSocketTransport.c [hunk 3]
```diff
@@ -0,0 +1,399 @@
+/*
+ * Copyright (c) 2018, 2024 SAP SE. All rights reserved.
+ * Copyright (c) 2018, 2024, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
```
Comment (RealCLanger) [inline src/jdk.jdwp.agent/share/native/libdt_filesocket/fileSocketTransport.c:4]: I think no Oracle copyright is necessary here.

**Review by RealCLanger — CHANGES_REQUESTED** [review]
I made some Copyright suggestions

File: make/modules/jdk.jdwp.agent/Lib.gmk [hunk 4]
```diff
@@ -47,6 +47,27 @@ TARGETS += $(BUILD_LIBDT_SOCKET)
 
 ################################################################################
 
+# SapMachine 2024-02-16: Add dt_filesocket implementation
+$(eval $(call SetupJdkLibrary, BUILD_LIBDT_FILESOCKET, \
+    NAME := dt_filesocket, \
+    OPTIMIZATION := LOW, \
```
[the remark this answers is not in the document]
Comment (schmelter-sap) [reply make/modules/jdk.jdwp.agent/Lib.gmk:53]: LOW is not really low (-O2 on gcc). It's the default for all created libs. Since the performance is dominated by the systemcalls anyways, going higher makes no sense.

File: src/jdk.jdwp.agent/share/native/libdt_filesocket/fileSocketTransport.c (hunk 3 above)
Comment (schmelter-sap) [reply src/jdk.jdwp.agent/share/native/libdt_filesocket/fileSocketTransport.c:4]: This file is based on socketTransport.c, so I've added this Oracle copyright.

File: make/modules/jdk.jdwp.agent/Lib.gmk (hunk 2 above)
Comment (schmelter-sap) [reply make/modules/jdk.jdwp.agent/Lib.gmk]: Done

File: src/jdk.jdwp.agent/share/native/libdt_filesocket/fileSocketTransport.h [hunk 5]
```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright (c) 2018, 2024 SAP SE. All rights reserved.
```
[the remark this answers is not in the document]
Comment (schmelter-sap) [reply src/jdk.jdwp.agent/share/native/libdt_filesocket/fileSocketTransport.h:2]: fileSocketTransport.c uses code from socketTransport.c, but this is not the case for this file.

File: src/jdk.jdwp.agent/share/native/libdt_filesocket/fileSocketTransport.h [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+#include "jni.h"
```
[the remark this answers is not in the document]
Comment (schmelter-sap) [reply src/jdk.jdwp.agent/share/native/libdt_filesocket/fileSocketTransport.h:26]: I fixed this.

File: src/jdk.jdwp.agent/unix/native/libdt_filesocket/fileSocketTransport_md.c [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+static int server_socket = -1;
+static int connection_socket = -1;
+
+static void closeSocket(int* socket) {
+    if (*socket != -1) {
+        int rv = -1;
+
+        do {
+            rv = close(*socket);
+        } while ((rv != 0) && (errno == EINTR));
```
[the remark this answers is not in the document]
Comment (schmelter-sap) [reply src/jdk.jdwp.agent/unix/native/libdt_filesocket/fileSocketTransport_md.c:60]: Good catch. I've changed the close implementation accordingly (and on Windows too).

**Comment by SapMachine** [comment]
Hello @schmelter-sap, this pull request fulfills all formal requirements.

File: src/jdk.jdwp.agent/share/native/libdt_filesocket/fileSocketTransport.c [hunk 8]
```diff
@@ -0,0 +1,399 @@
...
+            fileSocketTransport_logError("Could only write %d bytes instead of %d of the packet", (int) n, (int) (data_len - MAX_DATA_SIZE));
+            return JDWPTRANSPORT_ERROR_ILLEGAL_ARGUMENT;
+        }
+    }
+
+    return JDWPTRANSPORT_ERROR_NONE;
+}
+
+static jdwpTransportError JNICALL fileSocketTransport_GetLastError(jdwpTransportEnv* env, char** error) {
+    *error = (*callback->alloc)(sizeof(last_error));
```
[the remark this answers is not in the document]
Comment (schmelter-sap) [reply src/jdk.jdwp.agent/share/native/libdt_filesocket/fileSocketTransport.c:361]: Fixed

File: src/jdk.jdwp.agent/share/native/libdt_filesocket/fileSocketTransport.c [hunk 9]
```diff
@@ -0,0 +1,399 @@
...
+    if (initialized) {
+        return JNI_EEXIST;
+    }
+
+    initialized = JNI_TRUE;
+    jvm = vm;
+    callback = callbacks;
+
+    /* initialize interface table */
+    nif.GetCapabilities = fileSocketTransport_GetCapabilities;
```
[the remark this answers is not in the document]
Comment (schmelter-sap) [reply src/jdk.jdwp.agent/share/native/libdt_filesocket/fileSocketTransport.c:388]: It's the same in C. For a function func use can use 'func' itself or '&func' to get the address of the function.

File: src/jdk.jdwp.agent/share/native/libdt_filesocket/fileSocketTransport.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (schmelter-sap) [reply src/jdk.jdwp.agent/share/native/libdt_filesocket/fileSocketTransport.h]: No, that was an error. Should be fixed now.

**Comment by SapMachine** [comment]
Hello @schmelter-sap, this pull request fulfills all formal requirements.

**Comment by SapMachine** [comment]
Hello @schmelter-sap, this pull request fulfills all formal requirements.

File: make/common/modules/LauncherCommon.gmk [hunk 10]
```diff
@@ -122,9 +122,11 @@ define SetupBuildLauncherBody
     ifeq ($(STATIC_BUILD), true)
       $1_LDFLAGS += -exported_symbols_list \
               $(SUPPORT_OUTPUTDIR)/build-static/exported.symbols
+      # SapMachine 2024-02-16: Add dt_filesocket implementation
       $1_LIBS += \
           $$(shell $(FIND) $(SUPPORT_OUTPUTDIR)/modules_libs/java.base -name "*.a") \
           $(SUPPORT_OUTPUTDIR)/modules_libs/jdk.jdwp.agent/libdt_socket.a \
+          (SUPPORT_OUTPUTDIR)/modules_libs/jdk.jdwp.agent/libdt_filesocket.a \
```
Comment (RealCLanger) [inline make/common/modules/LauncherCommon.gmk:129]: Isn't a `$` here missing?

File: src/jdk.jdwp.agent/unix/native/libdt_filesocket/fileSocketTransport_md.c [hunk 11]
```diff
@@ -0,0 +1,270 @@
...
+#endif
+
+    /* Allow root too */
+    if (other_user != 0) {
+        if (other_user != geteuid()) {
+            fileSocketTransport_logError("Cannot allow user %d to connect to file socket %s of user %d",
+                (int)other_user, name, (int)geteuid());
+            fileSocketTransport_CloseImpl();
+        }
+        else if (other_group != getegid()) {
```
Comment (RealCLanger) [inline src/jdk.jdwp.agent/unix/native/libdt_filesocket/fileSocketTransport_md.c:234]: Formatting: should be `} else if `

File: src/jdk.jdwp.agent/windows/native/libdt_filesocket/fileSocketTransport_md.c [hunk 12]
```diff
@@ -0,0 +1,259 @@
...
+
+/* Make sure winsock is initialized on Windows */
+BOOL WINAPI DllMain(HINSTANCE hinst, DWORD reason, LPVOID reserved)
+{
+    WSADATA wsadata;
+
+    if ((reason == DLL_PROCESS_ATTACH) && (WSAStartup(MAKEWORD(2, 2), &wsadata) != 0)) {
+        return JNI_FALSE;
+    }
+    else if (reason == DLL_PROCESS_DETACH) {
```
Comment (RealCLanger) [inline src/jdk.jdwp.agent/windows/native/libdt_filesocket/fileSocketTransport_md.c:46]: Same here, formatting `} else if `

File: src/jdk.jdwp.agent/windows/native/libdt_filesocket/fileSocketTransport_md.c [hunk 13]
```diff
@@ -0,0 +1,259 @@
...
+static void registerFileToDelete(char const* name) {
+    if (file_to_delete_valid == 0) {
+        if ((name != NULL) && (strlen(name) + 1 <= sizeof(file_to_delete))) {
+            strcpy(file_to_delete, name);
+            MemoryBarrier();
+            file_to_delete_valid = 1;
+            MemoryBarrier();
+        }
+    }
+    else {
```
Comment (RealCLanger) [inline src/jdk.jdwp.agent/windows/native/libdt_filesocket/fileSocketTransport_md.c:83]: formatting

File: test/jdk/com/sun/jdi/FileSocketTransportTest.java [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+ */
+
+/**
+ * @test
+ * @summary Test basic functionality of the file socket debug transport.
+ *
+ * @author Ralf Schmelter
+ *
+ * @library /test/lib
+ * @run compile FileSocketTransportTest.java
```
Comment (RealCLanger) [inline test/jdk/com/sun/jdi/FileSocketTransportTest.java:31]: Is the `@compile` statement necessary ?

**Review by RealCLanger — CHANGES_REQUESTED** [review]
A few minor things/questions left

**Comment by SapMachine** [comment]
Hello @schmelter-sap, this pull request fulfills all formal requirements.

File: make/common/modules/LauncherCommon.gmk (hunk 10 above)
Comment (schmelter-sap) [reply make/common/modules/LauncherCommon.gmk]: I would guess so. Since it is only used in a static build. the error never materialized.

File: src/jdk.jdwp.agent/unix/native/libdt_filesocket/fileSocketTransport_md.c (hunk 11 above)
Comment (schmelter-sap) [reply src/jdk.jdwp.agent/unix/native/libdt_filesocket/fileSocketTransport_md.c]: Fixed

File: src/jdk.jdwp.agent/windows/native/libdt_filesocket/fileSocketTransport_md.c (hunk 13 above)
Comment (schmelter-sap) [reply src/jdk.jdwp.agent/windows/native/libdt_filesocket/fileSocketTransport_md.c]: Fixed

File: test/jdk/com/sun/jdi/FileSocketTransportTest.java (hunk 14 above)
Comment (schmelter-sap) [reply test/jdk/com/sun/jdi/FileSocketTransportTest.java]: Fixed

File: src/jdk.jdwp.agent/windows/native/libdt_filesocket/fileSocketTransport_md.c (hunk 12 above)
Comment (schmelter-sap) [reply src/jdk.jdwp.agent/windows/native/libdt_filesocket/fileSocketTransport_md.c]: Removed the else, since it is not needed at all.

**Review by RealCLanger — APPROVED** [review]
LGTM.

**Review by MBaesken — APPROVED** [review]
Thanks for doing the changes, LGTM now.
